feat(common): make date/number formatting functions public

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Feature
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
If you want to use the number, percent, currency and date pipes outside of your templates, you need to instantiate those pipes and manually transform the values

Issue Number: #20536


## What is the new behavior?
The utility functions `formatNumber`, `formatPercent`, `formatCurrency`, and `formatDate` used by the number, percent, currency and date pipes are now available for developers who want to use them outside of templates.


## Does this PR introduce a breaking change?
```
[x] No
```

mary-poppins commented:
You can preview aee9a42 at https://pr22372-aee9a42.ngbuilds.io/.

mary-poppins commented:
You can preview c4e518f at https://pr22372-c4e518f.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -40,25 +43,26 @@ enum TranslationType {
  * Transforms a date to a locale string based on a pattern and a timezone
  *
- * @internal
+ * @stable
```
needs more doc !

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
 }
+
+/**
+ * Transforms a value into a date
```
comment here do no say anything more than `toDate()`. 
Pls improve

vicb replied:
refine type also

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+    return new Date(parseFloat(value));
+  } else if (typeof value === 'string' && /^(\d{4}-\d{1,2}-\d{1,2})$/.test(value)) {
+    /**
+     * For ISO Strings without time the day, month and year must be extracted from the ISO String
```
move to API doc of the function ?

ocombe replied:
No, it's there to explain the method used to convert the ISO string to a date, but it has no value for people who use the function.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+  return date;
+}
+
+export function isoStringToDate(match: RegExpMatchArray): Date {
```
add a comment here pls

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -140,7 +140,7 @@ function formatNumber(
 /**
  * Formats a currency to a locale string
  *
```
please add docs

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -162,12 +162,12 @@ export function formatCurrency(
 /**
  * Formats a percentage to a locale string
  *
```
please add docs

vicb commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -120,6 +116,9 @@ export const ISO8601_DATE_REGEX =
  *   pipe needs to re-run (this is to avoid reformatting the date on every change detection run
  *   which would be an expensive operation).
  *
+ * If you want to use this pipe outside of a template, you can use the function
```
no you don't want to use the pipe outside of a template. Remove ?

vicb commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -31,6 +31,9 @@ import {invalidPipeArgumentError} from './invalid_pipe_argument_error';
  * For more information on the acceptable range for each of these numbers and other
  * details see your native internationalization library.
  *
+ * If you want to use this pipe outside of a template, you can use the function
```
same comment as other non-pipes

vicb commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -47,28 +48,44 @@ import localeAr from '@angular/common/locales/ar';
     describe('supports', () => {
-      it('should support date', () => { expect(() => pipe.transform(date)).not.toThrow(); });
+      it('should support date', () => {
+        expect(isDate(toDate(date))).toBeTruthy();
```
what is the intent here ?
Does this belong to `date_pipe_spec.ts` ?

vicb reviewed (changes requested):
Please
- Add or expand the API docs. Especially important for API docs. Internal functions should also have a short comment,
- refine date type from "any" to a|b|c,
- I was expected to see more tests moved to the new format spec. It doesn't make sense any more to have them in the pipes if pipes only forward to those functions (only a sanity check is ok for the pipes).

mary-poppins commented:
You can preview 0fe525a at https://pr22372-0fe525a.ngbuilds.io/.

mary-poppins commented:
You can preview d4d8baa at https://pr22372-d4d8baa.ngbuilds.io/.

mary-poppins commented:
You can preview a3fd774 at https://pr22372-a3fd774.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
  *
- * @internal
+ * Where:
+ * - `value` is a date object or a number (milliseconds since UTC epoch) or an ISO string
```
please update the type accordingly

vicb replied:
"is a `Date` or"

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
+ * (https://www.w3.org/TR/NOTE-datetime).
+ * - `format` indicates which date/time components to include. See {@link DatePipe} for a detailed
+ * description.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
no default

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
+ * default).
+ * - `timezone` to be used for formatting. It understands UTC/GMT and the continental US time zone
+ *  abbreviations, but for general use, use a time zone offset, for example,
+ *  `'+0430'` (4 hours, 30 minutes east of the Greenwich meridian)
```
remove "(4 hours, 30 minutes east of the Greenwich meridian)" - no added value and technically wrong GMT vs UTC

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
+ * - `timezone` to be used for formatting. It understands UTC/GMT and the continental US time zone
+ *  abbreviations, but for general use, use a time zone offset, for example,
+ *  `'+0430'` (4 hours, 30 minutes east of the Greenwich meridian)
+ *  If not specified, the local system timezone of the end-user's browser will be used.
```
"the local system timezone of the end-user's browser will be used." -> host system settings are used.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
+ *  `'+0430'` (4 hours, 30 minutes east of the Greenwich meridian)
+ *  If not specified, the local system timezone of the end-user's browser will be used.
+ *
+ * See {@link DatePipe} for more details about the parameters.
```
See {@link DatePipe} for more details

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
  */
-export function formatDate(date: Date, format: string, locale: string, timezone?: string): string {
+export function formatDate(value: any, format: string, locale: string, timezone?: string): string {
+  let date = toDate(value);
```
const

ocombe replied:
we change the value later if there is a timezone

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+
+    if (/^(\d{4}-\d{1,2}-\d{1,2})$/.test(value)) {
+      /**
+       * For ISO Strings without time the day, month and year must be extracted from the ISO String
```
move to the API doc of toDate

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
 }
+
+/**
+ * Transforms a value into a date. The supported values are dates, numbers, numeric strings
```
throw if not able to convert to a date

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+}
+
+/**
+ * This function converts a string matching the ISO8601_DATE_REGEX to a date.
```
"This function converts a string matching the ISO8601_DATE_REGEX to a date." 
What does this means ?
What about "Convert a date in ISO8601 to a `Date`" ?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+    tzMin = +(match[9] + match[11]);
+  }
+  dateSetter.call(date, +(match[1]), +(match[2]) - 1, +(match[3]));
+  const h = +(match[4] || '0') - tzHour;
```
+ -> Number, '0' -> 0

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -35,7 +35,7 @@ function strToNumber(value: number | string): number {
  * Transforms a number to a locale string based on a style and a format
  */
-function formatNumber(
+function formatNbToStr(
```
formatNbAsString ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -138,9 +138,22 @@ function formatNumber(
+ *
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
no default

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
+ * - `value` is a date object or a number (milliseconds since UTC epoch) or an ISO string
+ * (https://www.w3.org/TR/NOTE-datetime).
+ * - `format` indicates which date/time components to include. See {@link DatePipe} for a detailed
+ * description.
```
nit: I think aligning the text for continued line is easier to read

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
  *
- * @internal
+ * Where:
+ * - `value` is a number.
```
does not match the function !

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -138,9 +138,22 @@ function formatNumber(
+ * Use `currency` to format a number as currency.
+ *
+ * Where:
+ * - `value` is a number.
```
does not match the fn - number only seems to be the right thing

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
- * @internal
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
no default

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
+ * configurations are based on the active locale.
+ *
+ * Where:
+ * - `value` is a number.
```
again

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
+ * @description
+ *
+ * Formats a number as text. Group sizing and separator and other locale-specific
+ * configurations are based on the active locale.
```
active locale -> locale.

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
+ *
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
no default

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
+ * - `digitInfo` is a `string` which has a following format: <br>
+ *   <code>{minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}</code>.
+ *   - `minIntegerDigits` is the minimum number of integer digits to use. Defaults to `1`.
+ *   - `minFractionDigits` is the minimum number of digits after fraction. Defaults to `0`.
```
"after fraction" is it the correct terminology ?

vicb commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -0,0 +1,317 @@
+      it('should throw for objects', () => { expect(() => toDate({} as any)).toThrow(); });
+    });
+
+    describe('transform', () => {
```
transform ?

vicb commented on packages/common/test/i18n/format_number_spec.ts:
```diff
@@ -0,0 +1,130 @@
+import {formatCurrency, formatNumber, formatPercent, registerLocaleData} from '@angular/common';
+import {describe, expect, it} from '@angular/core/testing/src/testing_internal';
+
+{
```
remove this

vicb commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -0,0 +1,317 @@
+import localeTh from '@angular/common/locales/th';
+import {isDate, toDate, formatDate} from '@angular/common/src/i18n/format_date';
+
+{
```
remove

vicb commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -58,262 +46,23 @@ import localeAr from '@angular/common/locales/ar';
 
       it('should support ISO string',
-         () => expect(() => pipe.transform('2015-06-15T21:43:11Z')).not.toThrow());
+         () => { expect(() => pipe.transform('2015-06-15T21:43:11Z')).not.toThrow(); });
```
revert the diffs ?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
 }
+
+/**
+ * Converts a value to date
```
missing final .

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+}
+
+/**
+ * Convert a date in ISO8601 to a Date.
```
Convert*s*

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+
+/**
+ * Convert a date in ISO8601 to a Date.
+ * Used instead of Date.parse because of browser discrepancies.
```
quote `Date.parse`

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -138,9 +138,21 @@ function formatNumber(
+ * Use `currency` to format a number as currency.
+ *
+ * Where:
+ * - `value` is a number or numeric string.
```
wrong way ! -> number

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -160,28 +172,52 @@ export function formatCurrency(
+ * Where:
+ * - `value` is a number or numeric string.
+ * - `locale` is a `string` defining the locale to use.
+ * - `digitInfo` is a `string` which has a following format: <br>
```
add to all pipes or no pipe (other have `{@link ...}`)

mary-poppins commented:
You can preview da8586b at https://pr22372-da8586b.ngbuilds.io/.

wKoza commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -165,8 +184,10 @@ function padNumber(
+    strNum = '0' + strNum;
+  }
   if (trim) {
     strNum = strNum.substr(strNum.length - digits);
```
`strNum.length - digits` isn't always equal to `0`?

ocombe replied:
not always, we can have additional zeros

wKoza commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+ * Converts a value to date
+ *
+ * Supported input formats:
+ * - `Date`s
```
`Date`

mary-poppins commented:
You can preview 6d02858 at https://pr22372-6d02858.ngbuilds.io/.

ocombe committed:
feat(common): export functions to format numbers, percents, currencies & dates

The utility functions `formatNumber`, `formatPercent`, `formatCurrency`, and `formatDate` used by the number, percent, currency and date pipes are now available for developers who want to use them outside of templates.

Fixes #20536

mary-poppins commented:
You can preview dc77b4e at https://pr22372-dc77b4e.ngbuilds.io/.

IgorMinar commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -18,34 +18,19 @@ const DIGIT_CHAR = '#';
  */
-function formatNumber(
-    value: number | string, pattern: ParsedNumberFormat, locale: string, groupSymbol: NumberSymbol,
+function formatNbToLocaleString(
```
please avoid using custom abbreviations in non-local variables. formatNbToLocaleString -> formatNumberToLocaleString

anyone not familiar with our codebase that sees "Nb" will have to question what it means. This makes our code base harder to read and change by people not familiar with every change we make - e.g. our community, but also team members working on other parts of the code base that come across this code.

IgorMinar reviewed (approved):
the rest looks good. thanks

ocombe committed:
fixup!  feat(common): export functions to format numbers, percents, currencies & dates

vicb commented:
[presubmit](	http://test/OCL:186822308:BASE:186822499:1519424521742:7ea2c4fa
)

mary-poppins commented:
You can preview 8ce1f81 at https://pr22372-8ce1f81.ngbuilds.io/.
